Update source

aleksandrzajbel25182 opened the pull request.

aleksandrzajbel25182 committed:
Change docker-compose

aleksandrzajbel25182 committed:
Add dependencies and Datasource

aleksandrzajbel25182 committed:
Сreate models

aleksandrzajbel25182 committed:
Сreate a СRUD interface

aleksandrzajbel25182 committed:
Delete HelloServlet.java

aleksandrzajbel25182 committed:
Creating a test servlet

aleksandrzajbel25182 committed:
Refactor pom

aleksandrzajbel25182 committed:
Refactor model Currencies

aleksandrzajbel25182 committed:
Create Repository

aleksandrzajbel25182 committed:
Create DatabaseConfig

aleksandrzajbel25182 committed:
Refactor name

aleksandrzajbel25182 committed:
Create sql scripts database

aleksandrzajbel25182 committed:
Delete class

aleksandrzajbel25182 committed:
Add Lisner and HikarCp

Add .xml file config

aleksandrzajbel25182 committed:
Comment out the logic

aleksandrzajbel25182 committed:
Implement a connection pool

aleksandrzajbel25182 committed:
Logging dependence

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Do a search by currency id

Implement a currency search by id in the repository class.
Create a new servlet

aleksandrzajbel25182 committed:
Pretty Printing the Formatted JSON

aleksandrzajbel25182 committed:
Implement a servlet ExchangeRate

Implement the Exchange Rate repository and refine the model

aleksandrzajbel25182 committed:
add the ExchangeRateRepository to the listener

aleksandrzajbel25182 committed:
Rename ExchangeRatesRepository

aleksandrzajbel25182 committed:
CurrenciesRepository GetFindByCode

aleksandrzajbel25182 committed:
Implement a specific exchange rate servlet

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Add a method to create a generic model method

The createEntity method that creates an instance of the model class

aleksandrzajbel25182 committed:
Fix the typo in the variable names

aleksandrzajbel25182 committed:
Test Parsing

aleksandrzajbel25182 committed:
Gson Central Bank Parser

Make a Central Bank Currency model and create a custom deserialization

aleksandrzajbel25182 committed:
BugFix

aleksandrzajbel25182 committed:
Create Currencies

aleksandrzajbel25182 committed:
working with the parser

aleksandrzajbel25182 committed:
working with the parser 2.0

aleksandrzajbel25182 committed:
Updating the exchange rate

Updates are performed by parsing the quotes of the central bank

aleksandrzajbel25182 committed:
Connect the flyway plugin

Create a version 1 migration file

aleksandrzajbel25182 committed:
Remove unnecessary classes

aleksandrzajbel25182 committed:
Remove unnecessary classes

Create a web-app module and transfer everything related to the original
project.This is done to generate a new data update module in the
background.

aleksandrzajbel25182 committed:
Сhange the path of the .war file

After creating the web-app module, change the folder path.target in docker-compose

aleksandrzajbel25182 committed:
Create a new Update Source module

Create an abstract data source and a class to implement it.
Сreate a DTO template object
Create a class to update the database.

aleksandrzajbel25182 committed:
upload changes after the update

aleksandrzajbel25182 committed:
Add comments

shining-mind commented on .idea/dataSources.xml:
```diff
@@ -0,0 +1,18 @@
+      <driver-ref>postgresql</driver-ref>
+      <synchronize>true</synchronize>
+      <jdbc-driver>org.postgresql.Driver</jdbc-driver>
+      <jdbc-url>jdbc:postgresql://localhost:5432/currencyExchangedb</jdbc-url>
```
url лучше не хардкодить, так как в продакшне он скорее всего будет другим

shining-mind commented on .idea/dataSources.xml:
```diff
@@ -0,0 +1,18 @@
+      <jdbc-driver>org.postgresql.Driver</jdbc-driver>
+      <jdbc-url>jdbc:postgresql://localhost:5432/currencyExchangedb</jdbc-url>
+      <jdbc-additional-properties>
+        <property name="com.intellij.clouds.kubernetes.db.host.port" />
```
а причём тут `intellij.clouds.kubernetes`?

shining-mind commented on .idea/sqldialects.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="SqlDialectMappings">
+    <file url="file://$PROJECT_DIR$/web-app/src/main/resources/db/migration/V1__Create_Tables.sql" dialect="PostgreSQL" />
```
Немного непонятно как миграции связаны с SQL диалектами

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+
+  @Override
+  public ExchageRateDto get(LocalDate date) {
+    ExchageRateDto exchageRateDto  = new ExchageRateDto();
```
лишний пробел
```suggestion
    ExchageRateDto exchageRateDto = new ExchageRateDto();
```

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+    String month = date.format(monthFormatter);
+    int year = date.getYear();
+    final String URL = String.format("https://cbr.ru/scripts/XML_daily.asp?date_req=%s/%s/%d"
+        , day
```
немного странное форматирование, это так линтер отфартировал?

обычно пишут так:

```java
final String URL = String.format(
  "https://cbr.ru/scripts/XML_daily.asp?date_req=%s/%s/%d",
  day,
  month,
  year
);
```

shining-mind replied:
По крайне мере в обычном тексте
, мы же не ставим запятые в начале переноса - это странно выглядит

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+    for (int i = 0; i < valute.getLength(); i++) {
+      rates.add(createValute(valute.item(i)));
+    }
+    exchageRateDto.setDate(dateCBRF);
```
Непонятно, почему дата общая на все пары? А если у нас будет такой источник данных, где например разные пары имеют разную частоту обновления курса.

Например:
Делаем запрос на текущую дату, а он возвращает все актуальные курсы, но при этом у какой-то пары курс последний раз обновлялся неделю назад, но актуальный на текущую дату!

Потому у каждой пары должна быть своя дата

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+    String day = date.format(dayOfMonthFormatter);
+    String month = date.format(monthFormatter);
+    int year = date.getYear();
+    final String URL = String.format("https://cbr.ru/scripts/XML_daily.asp?date_req=%s/%s/%d"
```
Можно было не создавать переменную, а сразу пробросить её как аргумент функции

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+      DocumentBuilder documentBuilder =
+          DocumentBuilderFactory.newInstance().newDocumentBuilder();
+      document = documentBuilder.parse(url);
+    } catch (IOException e) {
```
А зачем перехватывать разные типы исключений и выкидывать их как RuntimeException? Не проще ли, чтобы он просто выкинул ошибку во вне и всё.

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -0,0 +1,116 @@
+
+
+  private Double toParseDouble(String str) {
+    String newStr =  str.replace(',', '.');
```
что-то с линтером - проверь, лишние пробелы пропускает

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,109 @@
+
+  public Updater() {
+
+    this.currenciesRepository = new CurrenciesRepository(null);
```
а для чего тут null?

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,109 @@
+
+    this.currenciesRepository = new CurrenciesRepository(null);
+    this.exchangeRateRepository = new ExchangeRateRepository(null);
+    this.cbrfSource = new CBRFSource();
```
  Updater должен работать с абстрактным источником, а не конкретным

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,109 @@
+
+
+    // Массив charcode
+    List<String> arrayCharCode = exchageRateDto.getCurrencies().stream()
```
Важно понимать, что тут происходит, на первый взгляд может показать, что у нас 3 раза выполняется обход всех элементов.
```
getCurrencies() -> List 
stream() -> Stream O(n) (создает новую структуру и перекладывает значения из исходного List) 
map() -> Stream O(n) (здесь 100% обходим все элементы)
collect() -> List O(n) (здесь опять перекладываем элементы в другую структуру)
```

Однако на деле stream API ленивый, вот подробное пояснение:

Когда вы преобразуете массив в поток в Java с помощью Arrays.stream(), операция сама по себе обычно легковесна и не включает немедленное копирование элементов массива в новую структуру. Вместо этого создается поток, который лениво обращается к элементам изначального массива по мере необходимости во время выполнения операций над потоком, таких как map, filter или collect.

Эта ленивая природа потоков означает, что операции над потоком не начинают выполняться до вызова терминальной операции (например, toArray(), collect(), forEach(), reduce() и т.д.). До терминальной операции все промежуточные операции (например, map) настраиваются в виде конвейера, и никакие фактические преобразования или обработка элементов не происходят. Это можно рассматривать как отложенное вычисление до тех пор, пока оно действительно не станет необходимым, что является основной особенностью API потоков Java, предназначенной для повышения эффективности и производительности, особенно при обработке больших наборов данных.

**Вот краткое описание:**
1. Создание потока: Arrays.stream(array) просто настраивает поток, который может получать доступ к элементам массивы. Этот шаг является "ленивым" в том смысле, что он не манипулирует элементами немедленно.

2. Промежуточные операции: Когда вы применяете операцию вроде map, она строит конвейер потока, но пока ничего не вычисляет. Она готовит трансформацию, но ждет выполнения.

3. Терминальные операции: Как только вы примените терминальную операцию, такую как toArray(), конвейер потока выполняет все настроенные операции последовательно над элементами массива, тем самым обрабатывая данные как указано.

Это ленивое поведение особенно полезно, когда вы соединяете несколько операций, так как оно помогает оптимизировать операции, уменьшая ненужные проходы по данным и управляя использованием памяти более эффективно. Например, если фильтрация значительно уменьшает набор данных, то последующие операции, такие как map, обрабатывают меньше элементов, что не было бы так оптимизировано, если бы все вычислялось сразу.

В заключение, преобразование массива в поток и применение map() не включает немедленную обработку массива или копирования его элементов, но создает рамки для обработки, которая выполняется только после вызова терминальной операции. Этот режим работы является фундаментальным для эффективности потоков в Java.

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,109 @@
+    // Массив charcode
+    List<String> arrayCharCode = exchageRateDto.getCurrencies().stream()
+        .map(CurrencyDto::getCharCode)
+        .collect(Collectors.toList());
```
Мне кажется `toList()` будет короче

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -0,0 +1,109 @@
+    var targetRate = exchangeRateRepository.findByDate(id, Date.valueOf(exchageRateDto.getDate()));
+
+    for (Integer id : currencies) {
+      var targerRate = exchangeRateRepository.findByDate(id,
```
Вот так делать плохо, если значения много их надо извлекать из БД все сразу. Используй оператор [IN](https://oracleplsql.ru/in-postgresql.html)

shining-mind commented on docker-compose.yml:
```diff
@@ -7,25 +7,28 @@ services:
     restart: always
     environment:
-      POSTGRES_USERNAME: root
+      POSTGRES_USER: postgres
```
пароли хранить в репе нельзя, используй ENV переменные https://stackoverflow.com/questions/29377853/how-can-i-use-environment-variables-in-docker-compose

shining-mind commented on web-app/src/main/resources/database.properties:
```diff
@@ -0,0 +1,4 @@
+DB_USER = postgres
+DB_PASSWORD = root
+DB_URL = jdbc:postgresql://pgsql:5432/currencyExchangedb
```
секретики не храним

shining-mind commented on web-app/src/main/java/exchangerate/utils/ConnectionPool.java:
```diff
@@ -0,0 +1,39 @@
+      Class.forName("org.postgresql.Driver");
+      config.setJdbcUrl("jdbc:postgresql://pgsql:5432/currencyExchangedb");
+      config.setUsername("postgres");
+      config.setPassword("root");
```
Надо брать из ENV секреты.

В итоге у тебя 10 мест, где объявлены секреты, надо использовать один источник, например .env файл

aleksandrzajbel25182 committed:
Hide the secrets

Refactoring pom.xml

aleksandrzajbel25182 committed:
Set it up .gitignore

aleksandrzajbel25182 committed:
Set it up .gitignore

aleksandrzajbel25182 committed:
Delete .env

aleksandrzajbel25182 committed:
Delete .idea repository

aleksandrzajbel25182 committed:
Store date properties with currency

aleksandrzajbel25182 committed:
Clean the Main class

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Refactoring and add new currencies with a list

aleksandrzajbel25182 committed:
Incomplete update algorithm

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
-    this.currenciesRepository = new CurrenciesRepository(null);
-    this.exchangeRateRepository = new ExchangeRateRepository(null);
-    this.cbrfSource = new CBRFSource();
+  public Updater(DataSource dataSource, CurrencyExchangeRateSource source, LocalDate date) {
```
dataSource - соединение с реляционной БД, а если у нас источник данных не БД?

Updater - это use-case приложения. По чистой архитектуре происходит нарушение.

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
+    // 1. Generate a charCode array
+    List<String> arrayCharCode = new ArrayList<>();
+    for (CurrencyDto curency : exchageRateDto.getCurrencies()) {
+      arrayCharCode.add(curency.getCharCode());
```
Здесь удобнее сразу создать HashMap,  так как он дальше будет нужен

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
-    // Из базы извлекаешь по всем полученным id курсы валют за дату
-    var targetRate = exchangeRateRepository.findByDate(id, Date.valueOf(exchageRateDto.getDate()));
+    // 2. For this array, request currency identifiers from the database
+    HashMap<Integer, String> idsDataBase = currenciesRepository.currencuIds(arrayCharCode);
```
```suggestion
    HashMap<Integer, String> idsDataBase = currenciesRepository.currencuIds(charCodeMap.keySet());
```

currencyIds - должен поддерживать интерфейс Collection или Iterable

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
-        //create
-      }
+    // 1. Generate a charCode array
+    List<String> arrayCharCode = new ArrayList<>();
```
```suggestion
    HashMap<String, CurrencyDto> charCodeMap = new HashMap();
```

shining-mind commented on web-app/src/main/java/exchangerate/repository/CurrenciesRepository.java:
```diff
@@ -67,20 +67,36 @@ public Currency findById(int id) {
   }
 
-  public Integer findCodeId(String hasCode) {
+  public HashMap<Integer, String> currencuIds(List<String> values) {
```
getCurrenciesIdByCode

shining-mind replied:
```suggestion
  public HashMap<String, Integer> currencuIds(List<String> values) {
```

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
-      if (targerRate.getRate().equals(hashMap.get(id).getValue())) {
+    // If the list from the database is empty, then enter it in the database
+    if (idsDataBase.isEmpty()) {
+      insertCurrency(exchageRateDto.getCurrencies());
```
```suggestion
      insertCurrency(exchageRateDto.getCurrencies());
      return;
```

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
+    // 3. Generate a HashMap id -> ObjectSource
+    currencyDtoHashMap = new HashMap<>();
+
+    for (Map.Entry<Integer, String> entry : idsDataBase.entrySet()) {
```
```java
exchangeRatesToUpdate = new ArrayList<Pair<Integer, ExchangeRateDto>>();
exchangeRatesToInsert = new ArrayList<ExchangeRateDto>();

for (Map.Entry<String, Integer> entry : idsDataBase.entrySet()) {
  ExchangeRateDto currency = charCodeMap.get(entry.getKey());

  if (currency != null) {
    exchangeRatesToUpdate.add(new Pair<>(entry.getValue(), currency));
  } else {
    exchangeRatesToInsert.add(currency);
  }
}

```

shining-mind replied:
Это алгоритм для ExchangeRate

shining-mind commented on UpdateSource/src/main/java/update/Source/CBRFSource.java:
```diff
@@ -24,19 +24,18 @@ public class CBRFSource implements CurrencyExchangeRateSource {
 
-
   @Override
   public ExchageRateDto get(LocalDate date) {
```
Источник должен возвращать массив пар валют

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
-      }
+    // 1. Generate a charCode array
+    List<String> arrayCharCode = new ArrayList<>();
+    for (CurrencyDto curency : exchageRateDto.getCurrencies()) {
```
```suggestion
    for (ExchangeRateDto curency : exchageRates) {
```

shining-mind commented on UpdateSource/src/main/java/update/Updater.java:
```diff
@@ -9,101 +9,94 @@
 
   public void updateExchageRate() {
 
+    exchageRateDto = source.get(date);
```
```suggestion
    List<ExchangeRateDto> exchageRates = source.get(date);
```

aleksandrzajbel25182 committed:
Migrate DB

aleksandrzajbel25182 committed:
Migrate DB

aleksandrzajbel25182 committed:
Fix dependency

aleksandrzajbel25182 committed:
Return an array of currency pairs in the source

One class is enough for a pair, the ruble acts as the currency in the CB RF

aleksandrzajbel25182 committed:
Make the Collection interface support

aleksandrzajbel25182 committed:
Not a perfect update code

aleksandrzajbel25182 committed:
Create a dto transformation in an entity

aleksandrzajbel25182 committed:
Fix sql queries

aleksandrzajbel25182 committed:
Delete the list package

aleksandrzajbel25182 committed:
Updating the exchange rate

Add an update to the currency table and correct the update algorithm

aleksandrzajbel25182 committed:
Refactoring

aleksandrzajbel25182 committed:
Add the main currency RUB

aleksandrzajbel25182 committed:
Fix message output

aleksandrzajbel25182 committed:
Fix the update algorithm

Additionally, make a request to the DB to get the currency exchange id
The request must also be present in the repository
